aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -0,0 +1,366 @@
+//          |                       /     mutations to a or b             |
+//          |                     /       are queued onto Alias           |
+//          |                   /                                         |
+//         \/                 /                                           \/
+// . - - - - - - - - - - - - - .                             . - - - - - - - - - - - - - - - .
+// |        TensorImpl         |                             |             TensorImpl        |
+// . - - - - - - - - - - - - - .                             . - - - - - - - - - - - - - - - .
+// |   value   |   storage     |                             |    storage    |     Value     |
+// . - - - - - - - - - - - - - .                             . - - - - - - - - - - - - - - - .
+//          |                                                             |
+//          |                                                             |
+//          |                                                             |
+//          |   In this picture the two tensor views their own storages,  |
+//          |   have their own storages, but backends like functorch      |
+//         \/   are allowed to re-alias underneath the pass               \/
+// . - - - - - - - - - - - - - .                             . - - - - - - - - - - - - - - - .
+// |    underyling_storage     |                             |      underyling_storage       |
+// . - - - - - - - - - - - - - .                             . - - - - - - - - - - - - - - - .
+//
+// See Note [Functionalization: Alias Removal Part 2] for more details on the mutation replay logic.
+void FunctionalTensorWrapper::set_view_meta(const Tensor& other, at::functionalization::ViewMeta meta) {
+    TORCH_INTERNAL_ASSERT(at::functionalization::impl::isFunctionalTensor(other));
+    auto other_impl = at::functionalization::impl::unsafeGetFunctionalWrapper(other);
+    TORCH_INTERNAL_ASSERT(this != other_impl);
+    TORCH_INTERNAL_ASSERT(!is_aliased())
+    auto other_storage_impl = static_cast<functionalization::FunctionalStorageImpl*>(other.unsafeGetTensorImpl()->unsafe_storage().unsafeGetStorageImpl());
+    if (!other_storage_impl->is_aliased()) {
+        // The original tensor wasn't a view - turn it into a (no-op) view.
+        other_storage_impl->set_alias(other_impl->value());
+    }
+
+    // Copy the original tensor's ViewMeta vector and push the current one.
+    if (other_impl->view_metas_.size() > 0) {
+        view_metas_ = other_impl->view_metas_;  // copy
+    }
+    view_metas_.push_back(meta);
+    storage_ = other_impl->storage_; // Ensure that storages are aliased properly.
+}
+
+// See Note [Functionalization Pass - Inplace View Ops]
+void FunctionalTensorWrapper::mutate_view_meta(at::functionalization::ViewMeta meta) {
+    auto self_impl_storage = static_cast<functionalization::FunctionalStorageImpl*>(unsafe_storage().unsafeGetStorageImpl());
```

Comment:
maybe make a helper for `static_cast<functionalization::FunctionalStorageImpl*>(unsafe_storage().unsafeGetStorageImpl())`